File: Graph/Python/Making A Large Island.py
# Problem link: https://leetcode.com/problems/making-a-large-island/description/
# Time: O(n * m)
# Space: O(n * m)
class Solution:
    def largestIsland(self, grid: List[List[int]]) -> int:
        result, rows, cols = 0, len(grid), len(grid[0])
        storeSize, parent = [[0] * cols for _ in range(rows)], [[-1] * (cols + 1) for _ in range(rows + 1)]
        directions = [[1, 0], [0, 1], [-1, 0], [0, -1]]

        def dfs(row, col, par):
            parent[row][col], size = par, 1
            
            for dr, dc in directions:
                nr, nc = dr + row, dc + col
                if 0 <= nr < rows and 0 <= nc < cols and grid[nr][nc] and parent[nr][nc] == -1:
                    size += dfs(row + dr, col + dc, par)
            
            return size

        for row in range(rows):
            for col in range(cols):
                if grid[row][col] and parent[row][col] == -1:
                    size = dfs(row, col, (row, col))
                    result, storeSize[row][col] = max(result, size), size
        
        for row in range(rows):
            for col in range(cols):
                if not grid[row][col]:
                    uniqueParents = set()

                    for dr, dc in directions:
                        if parent[row + dr][col + dc] != -1:
                            uniqueParents.add(parent[row + dr][col + dc])
                    
                    result = max(result, 1 + sum(storeSize[rw][cl] for rw, cl in uniqueParents))
        
        return result

File: Graph/Java/Making A Large Island.java
// Problem link: https://leetcode.com/problems/making-a-large-island/description/
// Time: O(n * m)
// Space: O(n * m)
class Solution {
    public int largestIsland(int[][] grid) {
        int result = 0, componentId = 1, rows = grid.length, cols = grid[0].length, size;
        int[] storeSize = new int[rows * cols + 1], parent[] = new int[rows][cols];
        int[][] directions = {{1, 0}, {0, 1}, {-1, 0}, {0, -1}};
        Set<Integer> uniqueComponents;

        for(int row = 0; row < rows; row++) {
            for(int col = 0; col < cols; col++) {
                if(grid[row][col] == 1 && parent[row][col] == 0) {
                    size = this.dfs(row, col, rows, cols, parent, grid, componentId, directions);
                    storeSize[componentId++] = size;
                    result = Math.max(result, size);
                }
            }
        }

        for(int row = 0; row < rows; row++) {
            for(int col = 0; col < cols; col++) {
                if(grid[row][col] == 0) {
                    uniqueComponents = new HashSet<>();
                    size = 0;

                    for(int[] dir : directions)
                        if(row + dir[0] >= 0 && row + dir[0] < rows && col + dir[1] >= 0 && col + dir[1] < cols)
                            uniqueComponents.add(parent[row + dir[0]][col + dir[1]]);

                    for(int component : uniqueComponents)
                        size += storeSize[component];

                    result = Math.max(result, 1 + size);
                }
            }
        }

        return result;
    }

    private int dfs(int row, int col, int rows, int cols, int[][] parent, int[][] grid, int componentId, int[][] directions) {
        parent[row][col] = componentId;
        int size = 1, nr, nc;

        for(int[] dir : directions) {
            nr = row + dir[0];
            nc = col + dir[1];

            if(nr >= 0 && nr < rows && nc >= 0 && nc < cols && grid[nr][nc] == 1 && parent[nr][nc] == 0) {
                size += dfs(nr, nc, rows, cols, parent, grid, componentId, directions);
            }
        }

        return size;
    }
}
